src/IRMatch.h

```diff
@@ -45,6 +47,1366 @@ bool expr_match(Expr pattern, Expr expr, std::map<std::string, Expr> &result);
 
 void expr_match_test();
 
+/** An alternative template-metaprogramming approach to expression
+ * matching. Potentially more efficient. We lift the expression
+ * pattern into a type, and then use force-inlined functions to
+ * generate efficient matching and reconstruction code for any
+ * pattern. Pattern elements are either one of the classes in the
+ * namespace IRMatcher, or are non-null Exprs (represented as
+ * BaseExprNode &).
+ *
+ * Pattern elements that are fully specified by their pattern can be
+ * built into an expression using the ::make method. Some patterns,
+ * such as a broadcast that matches any number of lanes, don't have
+ * enough information to recreate an Expr.
+ */
+namespace IRMatcher {
+
+/** To save stack space, the matcher objects are largely stateless and
+ * immutable. This state object is built up during matching and then
+ * consumed when constructing a replacement Expr.
+ */
+struct MatcherState {
+    const BaseExprNode *bindings[4] {nullptr, nullptr, nullptr, nullptr};
+
+    HALIDE_ALWAYS_INLINE
+    void set_binding(int i, const BaseExprNode &n) {
+        bindings[i] = &n;
+    }
+
+    HALIDE_ALWAYS_INLINE
+    bool is_bound(int i) const {
+        return bindings[i];
+    }
+
+    HALIDE_ALWAYS_INLINE
+    const BaseExprNode *get_binding(int i) const {
+        return bindings[i];
+    }
+
+    struct Const {
+        union {
+            double f;
+            uint64_t u;
+            int64_t i;
+        } val;
+        halide_type_t type;
+        uint32_t flags;
+    } bound_const[4];
+    uint32_t const_bound_mask = 0;
+
+    // flag masks for the flags field
+    static constexpr uint32_t signed_integer_overflow = 1;
+    static constexpr uint32_t indeterminate_expression = 2;
+
+    HALIDE_ALWAYS_INLINE
+    void set_bound_const(int i, double f, halide_type_t t) {
+        bound_const[i].val.f = f;
+        bound_const[i].type = t;
+        const_bound_mask |= (1 << i);
+    }
+
+    HALIDE_ALWAYS_INLINE
+    void set_bound_const(int i, uint64_t u, halide_type_t t) {
+        bound_const[i].val.u = u;
+        bound_const[i].type = t;
+        const_bound_mask |= (1 << i);
+    }
+
+    HALIDE_ALWAYS_INLINE
+    void set_bound_const(int i, int64_t s, halide_type_t t) {
+        bound_const[i].val.i = s;
+        bound_const[i].type = t;
+        const_bound_mask |= (1 << i);
+    }
+
+    HALIDE_ALWAYS_INLINE
+    const Const &get_bound_const(int i) const {
+        return bound_const[i];
+    }
+
+    HALIDE_ALWAYS_INLINE
+    bool is_bound_const(int i) const {
+        return const_bound_mask & (1 << i);
```

Comment:
return (const_bound_mask & (1 << i)) != 0;

(some compilers *will* bone you if you don't do this. ask me how I know)

Reply:
(though it's unlikely as long as the range of i is always < 4 ... still, good practice)